### ISSUE TITLE
Trailing commas after rest properties should be a syntax error

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/espree#310]
```js
require('espree').parse(
    'const { foo, ...bar, } = baz;',
    { ecmaVersion: 6, ecmaFeatures: { experimentalObjectRestSpread: true } }
);
```

Expected behavior: SyntaxError

Actual behavior: Successful parse

---

ESLint has received a large number of invalid `comma-dangle` bug reports indirectly caused by confusion from this bug. https://github.com/eslint/eslint/issues/7422, https://github.com/eslint/eslint/issues/7463, https://github.com/eslint/eslint/issues/7502, https://github.com/eslint/eslint/issues/7614, https://github.com/eslint/eslint/issues/7857, https://github.com/eslint/eslint/issues/8273, https://github.com/eslint/eslint/issues/8289.

That said, I'm not sure we should fix this until we do a major release. Babel actually the [same issue](https://github.com/babel/babylon/pull/149) at one point and they [fixed it](https://github.com/babel/babylon/releases/tag/v6.11.3), but they had to [revert it](https://github.com/babel/babylon/releases/tag/v6.11.4) almost immediately afterwards due to a lot of ecosystem breakage.

Still, I thought I'd make a tracking issue for it since it's incorrect behavior.

**Comment by ljharb** [comment]
False - this is object spread, not array spread. Trailing commas are *absolutely allowed* after object spread properties, because you can have more than one - whereas in array spread, only one, final rest property is allowed, hence no trailing comma.

This should be closed as invalid, and https://github.com/eslint/eslint/issues/7857 reopened.

**Comment by ljharb** [comment]
Oops, actually the example in this issue is invalid - but the example in the eslint issue is valid.

**Comment by hzoo** [comment]
👍 breaking mostly due to react-native and the fact that it compiles node_modules so users themselves didn't have to have trailing commas in their own codebase for the error to happen

**Comment by ljharb** [comment]
(See https://github.com/facebook/react-native/issues/10966 for the attempt to get the RN ecosystem to fix this problem)
